tests/factories.py

```diff
@@ -2,135 +2,159 @@
+
+    # Trying to add two DocumentURIs with the same claimant, uri, type and
+    # content_type to the db will crash. We use a sequence instead of something
+    # like FAKER.url() for claimant here so that never happens (unless you pass
+    # in your own claimant).
+    claimant = factory.Sequence(
+        lambda n: 'http://example.com/document_' + str(n) + '/')
+
+    uri = factory.LazyAttribute(lambda obj: obj.claimant)
+    type = factory.Iterator(['rel-alternate', 'rel-canonical', 'highwire-pdf',
+                             'dc-doi'])
+    content_type = factory.Iterator(['text/html', 'application/pdf',
+                                     'text/plain'])
+
+
+class Annotation(factory.alchemy.SQLAlchemyModelFactory):
+
+    class Meta:  # pylint: disable=no-init, old-style-class
+        model = models.Annotation
+        sqlalchemy_session = db.Session
+        force_flush = True  # Always flush the db to generate annotation.id.
+
+    tags = factory.LazyFunction(
+        lambda: FAKER.words(nb=random.randint(0, 5)))
+    target_uri = factory.Faker('uri')
+    text = factory.Faker('paragraph')
+    userid = factory.Faker('user_name')
+
+    @factory.lazy_attribute
+    def target_selectors(self):  # pylint: disable=no-self-use
+        return [
+            {
+                'endContainer': '/div[1]/article[1]/section[1]/div[1]/div[2]/div[1]',
+                'endOffset': 76,
+                'startContainer': '/div[1]/article[1]/section[1]/div[1]/div[2]/div[1]',
+                'startOffset': 0,
+                'type': 'RangeSelector'
+            },
+            {
+                'end': 362,
+                'start': 286,
+                'type': 'TextPositionSelector'
+            },
+            {
+                'exact': 'If you wish to install Hypothesis on your own site then head over to GitHub.',
+                'prefix': ' browser extension.\n            ',
+                'suffix': '\n          \n        \n      \n    ',
+                'type': 'TextQuoteSelector'
+            },
+        ]
+
+    @factory.post_generation
+    def make_document(self, create, extracted, **kwargs):
+        """Create associated document metadata for the annotation."""
+        # The metadata objects are going to be added to the db, so if we're not
+        # using the create strategy then simply don't make any.
+        if not create:
+            return
+
+        def document_uri_dict():
+            """
+            Return a randomly generated DocumentURI dict for this annotation.
+
+            This doesn't add anything to the database session yet.
+            """
+            document_uri = DocumentURI.build(claimant=self.target_uri,
+                                             uri=self.target_uri)
+            return dict(
+                claimant=document_uri.claimant,
+                uri=document_uri.uri,
+                type=document_uri.type,
+                content_type=document_uri.content_type,
+            )
```

Comment:
Here I want to be able to use all of factory_boy's goodness to randomly generate the values for the dict, and I think the only way to do that is to have a factory class to call. I could have written a `DocumentURIDict` factory class, but a `DocumentURI` factory seems more useful. So here I'm actually making an `h.api.models.DocumentURI` object, but not adding it to the db session, then returning a dict of some of its attributes.

It _might_ be possible to do something slightly nicer here, like have a `DocumentURI` factory class that _inherits_ from a `DocumentURIDict` factory without duplication.